drivers/gpio/gpio_silabs_siwx917_uulp.c

```diff
@@ -0,0 +1,205 @@
+		} else if (trig == GPIO_INT_TRIG_HIGH) {
+			flags = (mode == GPIO_INT_MODE_EDGE) ? SL_GPIO_INTERRUPT_RISE_EDGE
+							     : SL_GPIO_INTERRUPT_LEVEL_HIGH;
+		} else if (trig == GPIO_INT_TRIG_BOTH) {
+			/* SL_GPIO_INTERRUPT_RISE_FALL_EDGE would make more sense, but HAL
+			 * implementation is buggy.
+			 */
+			flags = SL_GPIO_INTERRUPT_RISE_EDGE | SL_GPIO_INTERRUPT_FALL_EDGE;
+		}
+
+		sl_si91x_gpio_configure_uulp_interrupt(flags, pin);
+	}
+	return 0;
+}
+
+static void gpio_siwx917_uulp_isr(const struct device *port)
+{
+	struct gpio_siwx917_uulp_data *data = port->data;
+	uint8_t pins = sl_si91x_gpio_get_uulp_interrupt_status();
+
+	sl_si91x_gpio_clear_uulp_interrupt(pins);
+
+	gpio_fire_callbacks(&data->callbacks, port, pins);
+}
+
+static const struct gpio_driver_api gpio_siwx917_uulp_driver_api = {
+	.pin_configure = gpio_siwx917_uulp_pin_configure,
+#ifdef CONFIG_GPIO_GET_CONFIG
+	.pin_get_config = NULL,
+#endif
+	.port_get_raw = gpio_siwx917_uulp_port_get,
+	.port_set_masked_raw = gpio_siwx917_uulp_port_set_masked,
+	.port_set_bits_raw = gpio_siwx917_uulp_port_set_bits,
+	.port_clear_bits_raw = gpio_siwx917_uulp_port_clear_bits,
+	.port_toggle_bits = gpio_siwx917_uulp_port_toggle_bits,
+	.pin_interrupt_configure = gpio_siwx917_uulp_interrupt_configure,
+	.manage_callback = gpio_siwx917_uulp_manage_callback,
+	.get_pending_int = NULL,
+#ifdef CONFIG_GPIO_GET_DIRECTION
+	.port_get_direction = NULL,
+#endif
+};
+
+#define GPIO_PORT_INIT(idx)                                                                        \
+	static int gpio_siwx917_init_uulp_##idx##_(const struct device *dev);                      \
+	static const struct gpio_siwx917_uulp_config gpio_siwx917_port_##idx##_config = {          \
+		.common =                                                                          \
+			{                                                                          \
+				.port_pin_mask = (gpio_port_pins_t)(-1),                           \
+			},                                                                         \
+	};                                                                                         \
+	static struct gpio_siwx917_uulp_data gpio_siwx917_port_##idx##_data;                       \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(idx, gpio_siwx917_init_uulp_##idx##_, NULL,                          \
+			      &gpio_siwx917_port_##idx##_data, &gpio_siwx917_port_##idx##_config,  \
+			      PRE_KERNEL_1, CONFIG_GPIO_INIT_PRIORITY,                             \
+			      &gpio_siwx917_uulp_driver_api);                                      \
+	static int gpio_siwx917_init_uulp_##idx##_(const struct device *dev)                       \
+	{                                                                                          \
+		GPIO_NPSS_GPIO_CONFIG_REG = 0;                                                     \
```

Comment:
Ideally, address of this register should come from the DT (note I think it is possible to specify several ranges in the `reg` property). That's said, I have to admit the registers looks scattered everywhere in the memory and the problem does not look simple :-(.

I would be also more comfortable with explicit `sys_write32()` rather than using `volatile` variables (we probably don't care on cortex-m4, but I believe `volatile` doesn't prevent for CPU reordering).

Reply:
Done.